<!DOCTYPE html>
<html lang="en">
<head>
<!-- 2018-01-17 Wed 15:23 -->
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<title>PIP calibration</title>
<meta name="generator" content="Org mode">
<meta name="author" content="Abhishek Sarkar">
<style type="text/css">
 <!--/*--><![CDATA[/*><!--*/
  .title  { text-align: center;
             margin-bottom: .2em; }
  .subtitle { text-align: center;
              font-size: medium;
              font-weight: bold;
              margin-top:0; }
  .todo   { font-family: monospace; color: red; }
  .done   { font-family: monospace; color: green; }
  .priority { font-family: monospace; color: orange; }
  .tag    { background-color: #eee; font-family: monospace;
            padding: 2px; font-size: 80%; font-weight: normal; }
  .timestamp { color: #bebebe; }
  .timestamp-kwd { color: #5f9ea0; }
  .org-right  { margin-left: auto; margin-right: 0px;  text-align: right; }
  .org-left   { margin-left: 0px;  margin-right: auto; text-align: left; }
  .org-center { margin-left: auto; margin-right: auto; text-align: center; }
  .underline { text-decoration: underline; }
  #postamble p, #preamble p { font-size: 90%; margin: .2em; }
  p.verse { margin-left: 3%; }
  pre {
    border: 1px solid #ccc;
    box-shadow: 3px 3px 3px #eee;
    padding: 8pt;
    font-family: monospace;
    overflow: auto;
    margin: 1.2em;
  }
  pre.src {
    position: relative;
    overflow: visible;
    padding-top: 1.2em;
  }
  pre.src:before {
    display: none;
    position: absolute;
    background-color: white;
    top: -10px;
    right: 10px;
    padding: 3px;
    border: 1px solid black;
  }
  pre.src:hover:before { display: inline;}
  /* Languages per Org manual */
  pre.src-asymptote:before { content: 'Asymptote'; }
  pre.src-awk:before { content: 'Awk'; }
  pre.src-C:before { content: 'C'; }
  /* pre.src-C++ doesn't work in CSS */
  pre.src-clojure:before { content: 'Clojure'; }
  pre.src-css:before { content: 'CSS'; }
  pre.src-D:before { content: 'D'; }
  pre.src-ditaa:before { content: 'ditaa'; }
  pre.src-dot:before { content: 'Graphviz'; }
  pre.src-calc:before { content: 'Emacs Calc'; }
  pre.src-emacs-lisp:before { content: 'Emacs Lisp'; }
  pre.src-fortran:before { content: 'Fortran'; }
  pre.src-gnuplot:before { content: 'gnuplot'; }
  pre.src-haskell:before { content: 'Haskell'; }
  pre.src-hledger:before { content: 'hledger'; }
  pre.src-java:before { content: 'Java'; }
  pre.src-js:before { content: 'Javascript'; }
  pre.src-latex:before { content: 'LaTeX'; }
  pre.src-ledger:before { content: 'Ledger'; }
  pre.src-lisp:before { content: 'Lisp'; }
  pre.src-lilypond:before { content: 'Lilypond'; }
  pre.src-lua:before { content: 'Lua'; }
  pre.src-matlab:before { content: 'MATLAB'; }
  pre.src-mscgen:before { content: 'Mscgen'; }
  pre.src-ocaml:before { content: 'Objective Caml'; }
  pre.src-octave:before { content: 'Octave'; }
  pre.src-org:before { content: 'Org mode'; }
  pre.src-oz:before { content: 'OZ'; }
  pre.src-plantuml:before { content: 'Plantuml'; }
  pre.src-processing:before { content: 'Processing.js'; }
  pre.src-python:before { content: 'Python'; }
  pre.src-R:before { content: 'R'; }
  pre.src-ruby:before { content: 'Ruby'; }
  pre.src-sass:before { content: 'Sass'; }
  pre.src-scheme:before { content: 'Scheme'; }
  pre.src-screen:before { content: 'Gnu Screen'; }
  pre.src-sed:before { content: 'Sed'; }
  pre.src-sh:before { content: 'shell'; }
  pre.src-sql:before { content: 'SQL'; }
  pre.src-sqlite:before { content: 'SQLite'; }
  /* additional languages in org.el's org-babel-load-languages alist */
  pre.src-forth:before { content: 'Forth'; }
  pre.src-io:before { content: 'IO'; }
  pre.src-J:before { content: 'J'; }
  pre.src-makefile:before { content: 'Makefile'; }
  pre.src-maxima:before { content: 'Maxima'; }
  pre.src-perl:before { content: 'Perl'; }
  pre.src-picolisp:before { content: 'Pico Lisp'; }
  pre.src-scala:before { content: 'Scala'; }
  pre.src-shell:before { content: 'Shell Script'; }
  pre.src-ebnf2ps:before { content: 'ebfn2ps'; }
  /* additional language identifiers per "defun org-babel-execute"
       in ob-*.el */
  pre.src-cpp:before  { content: 'C++'; }
  pre.src-abc:before  { content: 'ABC'; }
  pre.src-coq:before  { content: 'Coq'; }
  pre.src-groovy:before  { content: 'Groovy'; }
  /* additional language identifiers from org-babel-shell-names in
     ob-shell.el: ob-shell is the only babel language using a lambda to put
     the execution function name together. */
  pre.src-bash:before  { content: 'bash'; }
  pre.src-csh:before  { content: 'csh'; }
  pre.src-ash:before  { content: 'ash'; }
  pre.src-dash:before  { content: 'dash'; }
  pre.src-ksh:before  { content: 'ksh'; }
  pre.src-mksh:before  { content: 'mksh'; }
  pre.src-posh:before  { content: 'posh'; }
  /* Additional Emacs modes also supported by the LaTeX listings package */
  pre.src-ada:before { content: 'Ada'; }
  pre.src-asm:before { content: 'Assembler'; }
  pre.src-caml:before { content: 'Caml'; }
  pre.src-delphi:before { content: 'Delphi'; }
  pre.src-html:before { content: 'HTML'; }
  pre.src-idl:before { content: 'IDL'; }
  pre.src-mercury:before { content: 'Mercury'; }
  pre.src-metapost:before { content: 'MetaPost'; }
  pre.src-modula-2:before { content: 'Modula-2'; }
  pre.src-pascal:before { content: 'Pascal'; }
  pre.src-ps:before { content: 'PostScript'; }
  pre.src-prolog:before { content: 'Prolog'; }
  pre.src-simula:before { content: 'Simula'; }
  pre.src-tcl:before { content: 'tcl'; }
  pre.src-tex:before { content: 'TeX'; }
  pre.src-plain-tex:before { content: 'Plain TeX'; }
  pre.src-verilog:before { content: 'Verilog'; }
  pre.src-vhdl:before { content: 'VHDL'; }
  pre.src-xml:before { content: 'XML'; }
  pre.src-nxml:before { content: 'XML'; }
  /* add a generic configuration mode; LaTeX export needs an additional
     (add-to-list 'org-latex-listings-langs '(conf " ")) in .emacs */
  pre.src-conf:before { content: 'Configuration File'; }

  table { border-collapse:collapse; }
  caption.t-above { caption-side: top; }
  caption.t-bottom { caption-side: bottom; }
  td, th { vertical-align:top;  }
  th.org-right  { text-align: center;  }
  th.org-left   { text-align: center;   }
  th.org-center { text-align: center; }
  td.org-right  { text-align: right;  }
  td.org-left   { text-align: left;   }
  td.org-center { text-align: center; }
  dt { font-weight: bold; }
  .footpara { display: inline; }
  .footdef  { margin-bottom: 1em; }
  .figure { padding: 1em; }
  .figure p { text-align: center; }
  .inlinetask {
    padding: 10px;
    border: 2px solid gray;
    margin: 10px;
    background: #ffffcc;
  }
  #org-div-home-and-up
   { text-align: right; font-size: 70%; white-space: nowrap; }
  textarea { overflow-x: auto; }
  .linenr { font-size: smaller }
  .code-highlighted { background-color: #ffff00; }
  .org-info-js_info-navigation { border-style: none; }
  #org-info-js_console-label
    { font-size: 10px; font-weight: bold; white-space: nowrap; }
  .org-info-js_search-highlight
    { background-color: #ffff00; color: #000000; font-weight: bold; }
  .org-svg { width: 90%; }
  /*]]>*/-->
</style>
<link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" />
<link rel="stylesheet" type="text/css" href="https://www.pirilampo.org/styles/readtheorg/css/htmlize.css"/>
<style type="text/css">body {width: 60em; margin:auto} pre.src {overflow:auto}</style>
<script type="text/javascript">
/*
@licstart  The following is the entire license notice for the
JavaScript code in this tag.

Copyright (C) 2012-2017 Free Software Foundation, Inc.

The JavaScript code in this tag is free software: you can
redistribute it and/or modify it under the terms of the GNU
General Public License (GNU GPL) as published by the Free Software
Foundation, either version 3 of the License, or (at your option)
any later version.  The code is distributed WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE.  See the GNU GPL for more details.

As additional permission under GNU GPL version 3 section 7, you
may distribute non-source (e.g., minimized or compacted) forms of
that code without the copy of the GNU GPL normally required by
section 4, provided you include this license notice and a URL
through which recipients can access the Corresponding Source.


@licend  The above is the entire license notice
for the JavaScript code in this tag.
*/
<!--/*--><![CDATA[/*><!--*/
 function CodeHighlightOn(elem, id)
 {
   var target = document.getElementById(id);
   if(null != target) {
     elem.cacheClassElem = elem.className;
     elem.cacheClassTarget = target.className;
     target.className = "code-highlighted";
     elem.className   = "code-highlighted";
   }
 }
 function CodeHighlightOff(elem, id)
 {
   var target = document.getElementById(id);
   if(elem.cacheClassElem)
     elem.className = elem.cacheClassElem;
   if(elem.cacheClassTarget)
     target.className = elem.cacheClassTarget;
 }
/*]]>*///-->
</script>
<script type="text/x-mathjax-config">
    MathJax.Hub.Config({
        displayAlign: "center",
        displayIndent: "0em",

        "HTML-CSS": { scale: 100,
                        linebreaks: { automatic: "false" },
                        webFont: "TeX"
                       },
        SVG: {scale: 100,
              linebreaks: { automatic: "false" },
              font: "TeX"},
        NativeMML: {scale: 100},
        TeX: { equationNumbers: {autoNumber: "AMS"},
               MultLineWidth: "85%",
               TagSide: "right",
               TagIndent: ".8em"
             }
});
</script>
<script type="text/javascript"
        src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-AMS_HTML"></script>
</head>
<body>
<div id="content">
<h1 class="title">PIP calibration</h1>
<div id="table-of-contents">
<h2>Table of Contents</h2>
<div id="text-table-of-contents">
<ul>
<li><a href="#org068fe8b">Introduction</a></li>
<li><a href="#orgc62c868">One causal variant</a></li>
<li><a href="#orgdcb5f1c">Two causal variants</a></li>
</ul>
</div>
</div>

<div id="outline-container-org068fe8b" class="outline-2">
<h2 id="org068fe8b">Introduction</h2>
<div class="outline-text-2" id="text-org068fe8b">
<p>
Here we compare posterior inclusion probabilities (PIPs) computed using the
Sum of Single Effects (SSE) variational approximation against other methods:
</p>

<ul class="org-ul">
<li><a href="http://www.sciencedirect.com/science/article/pii/S0002929716300957?via%3Dihub">DAP</a> (2016). We used version 1.0, setting \(\phi = 0\) (no heterogeneity
between groups of samples) and \(\omega = 1\) (matching the
simulation). (<code>sse.wrapper.dap</code>)</li>
</ul>

<p>
The simulation generative model is as follows (<code>sse.evaluate._pip_calibration</code>):
</p>

<ol class="org-ol">
<li><p>
Sample a gene on chromosomes 1-22 which has at least 1,000 SNPs within 100
kilobases (respecting strand) of the gene start.
</p>

<div class="org-src-container">
<pre class="src src-ipython" id="org5de7e0c"><span class="org-variable-name">genotype_files</span> = [<span class="org-string">"/project/compbio/geuvadis/genotypes/GEUVADIS.chr{}.PH1PH2_465.IMPFRQFILT_BIALLELIC_PH.annotv2.genotypes.vcf.gz"</span>.<span class="org-builtin">format</span>(i) <span class="org-keyword">for</span> i <span class="org-keyword">in</span> <span class="org-builtin">range</span>(1, 23)]
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">def</span> <span class="org-function-name">filter_genes</span>(row, genotype_files, window, min_num_snps=1000):
  <span class="org-variable-name">window</span> = <span class="org-builtin">int</span>(window)
  <span class="org-variable-name">chrom</span> = <span class="org-builtin">int</span>(row[<span class="org-string">'chr'</span>][2:])
  <span class="org-keyword">if</span> row[<span class="org-string">'strand'</span>] == <span class="org-string">'+'</span>:
    <span class="org-variable-name">start</span> = row[<span class="org-string">'start'</span>] - window
    <span class="org-variable-name">end</span> = row[<span class="org-string">'start'</span>]
  <span class="org-keyword">else</span>:
    <span class="org-variable-name">start</span> = row[<span class="org-string">'start'</span>]
    <span class="org-variable-name">end</span> = row[<span class="org-string">'start'</span>] + window
  <span class="org-variable-name">count</span> = 0
  <span class="org-keyword">for</span> record <span class="org-keyword">in</span> tabix.<span class="org-builtin">open</span>(genotype_files[chrom - 1]).query(<span class="org-builtin">str</span>(chrom), start, end):
    <span class="org-variable-name">count</span> += 1
    <span class="org-keyword">if</span> count &gt; min_num_snps:
      <span class="org-keyword">return</span> <span class="org-constant">True</span>
  <span class="org-keyword">return</span> <span class="org-constant">False</span>
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython" id="org3f2d37e"><span class="org-variable-name">genes</span> = (pd.read_table(<span class="org-string">'/home/aksarkar/projects/singlecell-qtl/data/scqtl-genes.txt.gz'</span>)
         .set_index(<span class="org-string">'gene'</span>)
         .query(<span class="org-string">'source == "H. sapiens"'</span>)
         .query(<span class="org-string">'chr != "hsX"'</span>)
         .query(<span class="org-string">'chr != "hsY"'</span>)
         .query(<span class="org-string">'chr != "hsMT"'</span>))
<span class="org-variable-name">genes</span> = genes[genes.<span class="org-builtin">apply</span>(filter_genes, genotype_files=genotype_files, window=<span class="org-builtin">int</span>(1e5), axis=1)]
</pre>
</div>

<p>
Write out the filtered set of genes so we don't have to do this
long-running query more than once:
</p>

<div class="org-src-container">
<pre class="src src-ipython" id="org1df61b1">genes.to_csv(<span class="org-string">'/home/aksarkar/projects/sse/data/genes.txt.gz'</span>, sep=<span class="org-string">'\t'</span>, compression=<span class="org-string">'gzip'</span>)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython" id="orgde74c09"><span class="org-variable-name">genes</span> = pd.read_table(<span class="org-string">'/home/aksarkar/projects/sse/data/genes.txt.gz'</span>, index_col=0)
</pre>
</div></li>

<li>Extract genotypes of GEUVADIS individuals \(X\) from 100 kilobases
upstream (respecting strand) of the gene start (<code>sse.evaluate.read_vcf</code>).</li>
<li value="3">Estimate MAF \(f_j\) from \(X\) (<code>sse.simulation.Simulation.estimate_maf</code>)</li>
<li>Fix proportion of variance explained \(h^2 = 0.15\) (mean
<i>cis</i>-heritability of gene expression across all genes; <a href="https://www.nature.com/articles/ng.3506">Gusev et al 2016</a>,
<a href="https://www.nature.com/articles/ng.3506">Wheeler et al 2016</a>)</li>
<li>Sample causal effects from \(\beta_j \sim N(0, 1)\)
(<code>sse.simulation.Simulation.sample_effects</code>).</li>
<li>Compute genetic variance \(V_g = \sum_j 2 f_j (1 - f_j) \beta_j^2\)</li>
<li>Sample residuals \(\epsilon_i \sim N(0, V_g (1 / h^2 - 1))\)</li>
<li>Compute phenotypes \(y_i = X_i\beta + \epsilon_i\) (<code>sse.simulation.Simulation.compute_liabilities</code>)</li>
</ol>
</div>
</div>

<div id="outline-container-orgc62c868" class="outline-2">
<h2 id="orgc62c868">One causal variant</h2>
<div class="outline-text-2" id="text-orgc62c868">
<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">one_causal</span> = sse.evaluate.pip_calibration(
  genes=genes,
  genotype_files=genotype_files,
  num_genes=100,
  num_trials=1,
  num_effects=10,
  seed=0)
</pre>
</div>

<p>
Tabulate the actual number of variants used in the simulation:
</p>

<div class="org-src-container">
<pre class="src src-ipython">one_causal.groupby(level=0).<span class="org-builtin">apply</span>(<span class="org-keyword">lambda</span> x: x.iloc[0][<span class="org-string">'num_snps'</span>]).describe()
</pre>
</div>

<pre class="example">
count     100.000000
mean     1381.850000
std       216.044672
min      1014.000000
25%      1230.750000
50%      1376.500000
75%      1497.000000
max      2320.000000
dtype: float64
</pre>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">del</span> one_causal[<span class="org-string">'num_snps'</span>]
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()
plt.boxplot(one_causal.fillna(0).values, labels=one_causal.columns)
plt.xlabel(<span class="org-string">'Method'</span>)
plt.ylabel(<span class="org-string">'Maximum absolute difference in PIP'</span>)
plt.gcf()
</pre>
</div>


<div class="figure">
<p><img src="one-causal-variant-max-error.png" alt="one-causal-variant-max-error.png">
</p>
</div>
</div>
</div>

<div id="outline-container-orgdcb5f1c" class="outline-2">
<h2 id="orgdcb5f1c">Two causal variants</h2>
<div class="outline-text-2" id="text-orgdcb5f1c">
<p>
This takes ~4 hours to run.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">two_causal</span> = sse.evaluate.pip_calibration(
  genes=genes,
  genotype_files=genotype_files,
  num_genes=100,
  num_trials=1,
  num_causal=2,
  num_effects=10,
  seed=0)
two_causal.to_csv(<span class="org-string">'/home/aksarkar/projects/sse/data/two-causal-max-error.txt.gz'</span>, sep=<span class="org-string">'\t'</span>, compression=<span class="org-string">'gzip'</span>)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-emacs-lisp">(org-tangle-buffer)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-sh">sbatch --partition=<span class="org-string">"broadwl"</span> --mem=4G --time=200 --job-name=<span class="org-string">"sse"</span>
<span class="org-comment-delimiter">#</span><span class="org-comment">!/bin/bash</span>
<span class="org-builtin">source</span> activate sse
python pip.py
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">two_causal</span> = pd.read_table(<span class="org-string">'/home/aksarkar/projects/sse/data/two-causal-max-error.txt.gz'</span>, index_col=[0, 1])
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">del</span> two_causal[<span class="org-string">'num_snps'</span>]
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()
plt.boxplot(two_causal.fillna(0).values, labels=two_causal.columns)
plt.xlabel(<span class="org-string">'Method'</span>)
plt.ylabel(<span class="org-string">'Maximum absolute difference in PIP'</span>)
plt.gcf()
</pre>
</div>


<div class="figure">
<p><img src="two-causal-variants-max-error.png" alt="two-causal-variants-max-error.png">
</p>
</div>
</div>
</div>
</div>
<div id="postamble" class="status">
<p class="author">Author: Abhishek Sarkar</p>
<p class="date">Created: 2018-01-17 Wed 15:23</p>
<p class="validation"><a href="http://validator.w3.org/check?uri=referer">Validate</a></p>
</div>
</body>
</html>
