<!DOCTYPE html>
<html lang="en">
<head>
<!-- 2017-12-02 Sat 00:10 -->
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<title>PIP calibration</title>
<meta name="generator" content="Org mode">
<meta name="author" content="Abhishek Sarkar">
<style type="text/css">
 <!--/*--><![CDATA[/*><!--*/
  .title  { text-align: center;
             margin-bottom: .2em; }
  .subtitle { text-align: center;
              font-size: medium;
              font-weight: bold;
              margin-top:0; }
  .todo   { font-family: monospace; color: red; }
  .done   { font-family: monospace; color: green; }
  .priority { font-family: monospace; color: orange; }
  .tag    { background-color: #eee; font-family: monospace;
            padding: 2px; font-size: 80%; font-weight: normal; }
  .timestamp { color: #bebebe; }
  .timestamp-kwd { color: #5f9ea0; }
  .org-right  { margin-left: auto; margin-right: 0px;  text-align: right; }
  .org-left   { margin-left: 0px;  margin-right: auto; text-align: left; }
  .org-center { margin-left: auto; margin-right: auto; text-align: center; }
  .underline { text-decoration: underline; }
  #postamble p, #preamble p { font-size: 90%; margin: .2em; }
  p.verse { margin-left: 3%; }
  pre {
    border: 1px solid #ccc;
    box-shadow: 3px 3px 3px #eee;
    padding: 8pt;
    font-family: monospace;
    overflow: auto;
    margin: 1.2em;
  }
  pre.src {
    position: relative;
    overflow: visible;
    padding-top: 1.2em;
  }
  pre.src:before {
    display: none;
    position: absolute;
    background-color: white;
    top: -10px;
    right: 10px;
    padding: 3px;
    border: 1px solid black;
  }
  pre.src:hover:before { display: inline;}
  /* Languages per Org manual */
  pre.src-asymptote:before { content: 'Asymptote'; }
  pre.src-awk:before { content: 'Awk'; }
  pre.src-C:before { content: 'C'; }
  /* pre.src-C++ doesn't work in CSS */
  pre.src-clojure:before { content: 'Clojure'; }
  pre.src-css:before { content: 'CSS'; }
  pre.src-D:before { content: 'D'; }
  pre.src-ditaa:before { content: 'ditaa'; }
  pre.src-dot:before { content: 'Graphviz'; }
  pre.src-calc:before { content: 'Emacs Calc'; }
  pre.src-emacs-lisp:before { content: 'Emacs Lisp'; }
  pre.src-fortran:before { content: 'Fortran'; }
  pre.src-gnuplot:before { content: 'gnuplot'; }
  pre.src-haskell:before { content: 'Haskell'; }
  pre.src-hledger:before { content: 'hledger'; }
  pre.src-java:before { content: 'Java'; }
  pre.src-js:before { content: 'Javascript'; }
  pre.src-latex:before { content: 'LaTeX'; }
  pre.src-ledger:before { content: 'Ledger'; }
  pre.src-lisp:before { content: 'Lisp'; }
  pre.src-lilypond:before { content: 'Lilypond'; }
  pre.src-lua:before { content: 'Lua'; }
  pre.src-matlab:before { content: 'MATLAB'; }
  pre.src-mscgen:before { content: 'Mscgen'; }
  pre.src-ocaml:before { content: 'Objective Caml'; }
  pre.src-octave:before { content: 'Octave'; }
  pre.src-org:before { content: 'Org mode'; }
  pre.src-oz:before { content: 'OZ'; }
  pre.src-plantuml:before { content: 'Plantuml'; }
  pre.src-processing:before { content: 'Processing.js'; }
  pre.src-python:before { content: 'Python'; }
  pre.src-R:before { content: 'R'; }
  pre.src-ruby:before { content: 'Ruby'; }
  pre.src-sass:before { content: 'Sass'; }
  pre.src-scheme:before { content: 'Scheme'; }
  pre.src-screen:before { content: 'Gnu Screen'; }
  pre.src-sed:before { content: 'Sed'; }
  pre.src-sh:before { content: 'shell'; }
  pre.src-sql:before { content: 'SQL'; }
  pre.src-sqlite:before { content: 'SQLite'; }
  /* additional languages in org.el's org-babel-load-languages alist */
  pre.src-forth:before { content: 'Forth'; }
  pre.src-io:before { content: 'IO'; }
  pre.src-J:before { content: 'J'; }
  pre.src-makefile:before { content: 'Makefile'; }
  pre.src-maxima:before { content: 'Maxima'; }
  pre.src-perl:before { content: 'Perl'; }
  pre.src-picolisp:before { content: 'Pico Lisp'; }
  pre.src-scala:before { content: 'Scala'; }
  pre.src-shell:before { content: 'Shell Script'; }
  pre.src-ebnf2ps:before { content: 'ebfn2ps'; }
  /* additional language identifiers per "defun org-babel-execute"
       in ob-*.el */
  pre.src-cpp:before  { content: 'C++'; }
  pre.src-abc:before  { content: 'ABC'; }
  pre.src-coq:before  { content: 'Coq'; }
  pre.src-groovy:before  { content: 'Groovy'; }
  /* additional language identifiers from org-babel-shell-names in
     ob-shell.el: ob-shell is the only babel language using a lambda to put
     the execution function name together. */
  pre.src-bash:before  { content: 'bash'; }
  pre.src-csh:before  { content: 'csh'; }
  pre.src-ash:before  { content: 'ash'; }
  pre.src-dash:before  { content: 'dash'; }
  pre.src-ksh:before  { content: 'ksh'; }
  pre.src-mksh:before  { content: 'mksh'; }
  pre.src-posh:before  { content: 'posh'; }
  /* Additional Emacs modes also supported by the LaTeX listings package */
  pre.src-ada:before { content: 'Ada'; }
  pre.src-asm:before { content: 'Assembler'; }
  pre.src-caml:before { content: 'Caml'; }
  pre.src-delphi:before { content: 'Delphi'; }
  pre.src-html:before { content: 'HTML'; }
  pre.src-idl:before { content: 'IDL'; }
  pre.src-mercury:before { content: 'Mercury'; }
  pre.src-metapost:before { content: 'MetaPost'; }
  pre.src-modula-2:before { content: 'Modula-2'; }
  pre.src-pascal:before { content: 'Pascal'; }
  pre.src-ps:before { content: 'PostScript'; }
  pre.src-prolog:before { content: 'Prolog'; }
  pre.src-simula:before { content: 'Simula'; }
  pre.src-tcl:before { content: 'tcl'; }
  pre.src-tex:before { content: 'TeX'; }
  pre.src-plain-tex:before { content: 'Plain TeX'; }
  pre.src-verilog:before { content: 'Verilog'; }
  pre.src-vhdl:before { content: 'VHDL'; }
  pre.src-xml:before { content: 'XML'; }
  pre.src-nxml:before { content: 'XML'; }
  /* add a generic configuration mode; LaTeX export needs an additional
     (add-to-list 'org-latex-listings-langs '(conf " ")) in .emacs */
  pre.src-conf:before { content: 'Configuration File'; }

  table { border-collapse:collapse; }
  caption.t-above { caption-side: top; }
  caption.t-bottom { caption-side: bottom; }
  td, th { vertical-align:top;  }
  th.org-right  { text-align: center;  }
  th.org-left   { text-align: center;   }
  th.org-center { text-align: center; }
  td.org-right  { text-align: right;  }
  td.org-left   { text-align: left;   }
  td.org-center { text-align: center; }
  dt { font-weight: bold; }
  .footpara { display: inline; }
  .footdef  { margin-bottom: 1em; }
  .figure { padding: 1em; }
  .figure p { text-align: center; }
  .inlinetask {
    padding: 10px;
    border: 2px solid gray;
    margin: 10px;
    background: #ffffcc;
  }
  #org-div-home-and-up
   { text-align: right; font-size: 70%; white-space: nowrap; }
  textarea { overflow-x: auto; }
  .linenr { font-size: smaller }
  .code-highlighted { background-color: #ffff00; }
  .org-info-js_info-navigation { border-style: none; }
  #org-info-js_console-label
    { font-size: 10px; font-weight: bold; white-space: nowrap; }
  .org-info-js_search-highlight
    { background-color: #ffff00; color: #000000; font-weight: bold; }
  .org-svg { width: 90%; }
  /*]]>*/-->
</style>
<link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" />
<link rel="stylesheet" type="text/css" href="https://www.pirilampo.org/styles/readtheorg/css/htmlize.css"/>
<style type="text/css">body {width: 60em; margin:auto} pre.src {overflow:auto}</style>
<script type="text/javascript">
/*
@licstart  The following is the entire license notice for the
JavaScript code in this tag.

Copyright (C) 2012-2017 Free Software Foundation, Inc.

The JavaScript code in this tag is free software: you can
redistribute it and/or modify it under the terms of the GNU
General Public License (GNU GPL) as published by the Free Software
Foundation, either version 3 of the License, or (at your option)
any later version.  The code is distributed WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE.  See the GNU GPL for more details.

As additional permission under GNU GPL version 3 section 7, you
may distribute non-source (e.g., minimized or compacted) forms of
that code without the copy of the GNU GPL normally required by
section 4, provided you include this license notice and a URL
through which recipients can access the Corresponding Source.


@licend  The above is the entire license notice
for the JavaScript code in this tag.
*/
<!--/*--><![CDATA[/*><!--*/
 function CodeHighlightOn(elem, id)
 {
   var target = document.getElementById(id);
   if(null != target) {
     elem.cacheClassElem = elem.className;
     elem.cacheClassTarget = target.className;
     target.className = "code-highlighted";
     elem.className   = "code-highlighted";
   }
 }
 function CodeHighlightOff(elem, id)
 {
   var target = document.getElementById(id);
   if(elem.cacheClassElem)
     elem.className = elem.cacheClassElem;
   if(elem.cacheClassTarget)
     target.className = elem.cacheClassTarget;
 }
/*]]>*///-->
</script>
<script type="text/x-mathjax-config">
    MathJax.Hub.Config({
        displayAlign: "center",
        displayIndent: "0em",

        "HTML-CSS": { scale: 100,
                        linebreaks: { automatic: "false" },
                        webFont: "TeX"
                       },
        SVG: {scale: 100,
              linebreaks: { automatic: "false" },
              font: "TeX"},
        NativeMML: {scale: 100},
        TeX: { equationNumbers: {autoNumber: "AMS"},
               MultLineWidth: "85%",
               TagSide: "right",
               TagIndent: ".8em"
             }
});
</script>
<script type="text/javascript"
        src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-AMS_HTML"></script>
</head>
<body>
<div id="content">
<h1 class="title">PIP calibration</h1>
<div id="table-of-contents">
<h2>Table of Contents</h2>
<div id="text-table-of-contents">
<ul>
<li><a href="#org068fe8b">Introduction</a></li>
<li><a href="#orgc62c868">One causal variant</a></li>
</ul>
</div>
</div>

<div id="outline-container-org068fe8b" class="outline-2">
<h2 id="org068fe8b">Introduction</h2>
<div class="outline-text-2" id="text-org068fe8b">
<p>
Here we estimate Spearman correlation between posterior inclusion
probabilities (PIPs) computed using the Sum of Single Effects (SSE)
variational approximation against other methods:
</p>

<ul class="org-ul">
<li><a href="https://academic.oup.com/bioinformatics/article-lookup/doi/10.1093/bioinformatics/btw018">FINEMAP</a> (2016). We used version 1.1, setting the prior on the effect size
scale to \(N(0, 1)\) (matching the simulation). We used defaults for all
other settings. (<code>sse.wrapper.finemap</code>)</li>

<li><a href="http://www.sciencedirect.com/science/article/pii/S0002929716300957?via%3Dihub">DAP</a> (2016). We used version 1.0, setting \(\phi = 0\) (no heterogeneity
between groups of samples) and \(\omega = 1\) (matching the
simulation). (<code>sse.wrapper.dap</code>)</li>
</ul>

<p>
The simulation generative model is as follows:
</p>

<ol class="org-ol">
<li><p>
Sample a gene on chromosomes 1-22
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">genes</span> = (pd.read_table(<span class="org-string">'/home/aksarkar/projects/singlecell-qtl/data/scqtl-genes.txt.gz'</span>)
         .set_index(<span class="org-string">'gene'</span>)
         .query(<span class="org-string">'source == "H. sapiens" and chr != "hsX" and chr != "hsY"'</span>))
</pre>
</div></li>

<li><p>
Extract genotypes of GEUVADIS individuals \(X\) from 100 kilobases
upstream (respecting strand) of the gene start (<code>sse.evaluate.read_vcf</code>)
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">genotype_files</span> = [<span class="org-string">"/project/compbio/geuvadis/genotypes/GEUVADIS.chr{}.PH1PH2_465.IMPFRQFILT_BIALLELIC_PH.annotv2.genotypes.vcf.gz"</span>.<span class="org-builtin">format</span>(i) <span class="org-keyword">for</span> i <span class="org-keyword">in</span> <span class="org-builtin">range</span>(1, 23)]
</pre>
</div></li>

<li value="3">Estimate MAF \(f_j\) from \(X\) (<code>sse.simulation.Simulation.estimate_maf</code>)</li>
<li>Fix proportion of variance explained \(h^2 = 0.15\) (mean
<i>cis</i>-heritability of gene expression across all genes; <a href="https://www.nature.com/articles/ng.3506">Gusev et al 2016</a>,
<a href="https://www.nature.com/articles/ng.3506">Wheeler et al 2016</a>)</li>
<li>Sample causal effects from \(\beta_j \sim N(0, 1)\)
(<code>sse.simulation.Simulation.sample_effects</code>). To ensure the methods can
detect the causal variant, we only sample from among variants with minor
allele frequency \(> 0.25\). (<code>sse.evaluate._generate_pheno</code>)</li>
<li>Compute genetic variance \(V_g = \sum_j 2 f_j (1 - f_j) \beta_j^2\)</li>
<li>Sample residuals \(\epsilon_i \sim N(0, V_g (1 / h^2 - 1))\)</li>
<li>Compute phenotypes \(y_i = X_i\beta + \epsilon_i\) (<code>sse.simulation.Simulation.compute_liabilities</code>)</li>
</ol>
</div>
</div>

<div id="outline-container-orgc62c868" class="outline-2">
<h2 id="orgc62c868">One causal variant</h2>
<div class="outline-text-2" id="text-orgc62c868">
<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">one_causal</span> = sse.evaluate.pip_calibration(
  genes=genes,
  genotype_files=genotype_files,
  num_genes=1,
  num_trials=10,
  num_effects=10,
  max_epochs=100,
  seed=0)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()
plt.boxplot(one_causal.values, labels=one_causal.columns)
plt.xlabel(<span class="org-string">'Method'</span>)
plt.ylabel(<span class="org-string">'Spearman correlation with SSE PIP'</span>)
plt.gcf()
</pre>
</div>


<div class="figure">
<p><img src="one-causal-variant-spearman.png" alt="one-causal-variant-spearman.png">
</p>
</div>

<p>
Investigate what's going on with FINEMAP:
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">x</span>, <span class="org-variable-name">y</span>, <span class="org-variable-name">s</span> = sse.evaluate._generate_pheno(0, genes.loc[<span class="org-string">"ENSG00000143641"</span>], genotype_files, 1, <span class="org-builtin">int</span>(1e5))
<span class="org-variable-name">pip</span> = (sse.model.GaussianSSE()
       .fit(x, y)
       .pip_df
       .agg(np.<span class="org-builtin">sum</span>, axis=1)
       .to_frame()
       .merge(sse.wrapper.finemap(x, y), left_index=<span class="org-constant">True</span>, right_index=<span class="org-constant">True</span>)
       .rename(columns={0: <span class="org-string">'sse_pip'</span>, <span class="org-string">'pip'</span>: <span class="org-string">'finemap_pip'</span>}))
<span class="org-variable-name">pip</span>[<span class="org-string">'causal'</span>] = s.theta != 0
<span class="org-variable-name">pip</span>[<span class="org-string">'true_effect'</span>] = s.theta
</pre>
</div>

<p>
SSE correctly recovers the causal variant, although the PIPs suggest that
the simulated problem might be difficult for any method.
</p>

<div class="org-src-container">
<pre class="src src-ipython">pip.sort_values(by=<span class="org-string">'sse_pip'</span>, ascending=<span class="org-constant">False</span>).head()
</pre>
</div>

<pre class="example">
        sse_pip  index  finemap_pip  log10bf  causal  true_effect
snp1278  0.264491   1279       0.1968   2.3662    True     1.031421
snp1232  0.225833   1233       0.1843   2.3312   False     0.000000
snp714   0.114354    715       0.1195   2.1096   False     0.000000
snp980   0.085730    981       0.0867   1.9547   False     0.000000
snp1122  0.071195   1123       0.0548   1.7401   False     0.000000
</pre>

<p>
FINEMAP assigns stronger PIP to non-causal variants.
</p>

<div class="org-src-container">
<pre class="src src-ipython">pip.sort_values(by=<span class="org-string">'log10bf'</span>, ascending=<span class="org-constant">False</span>).head()
</pre>
</div>

<pre class="example">
            sse_pip  index  finemap_pip  log10bf  causal  true_effect
snp1398  2.676482e-05   1399       1.0000  12.9767   False     0.000000
snp1404  1.135552e-08   1405       0.9998   6.6873   False     0.000000
snp676   7.362895e-14    677       0.8887   3.8795   False     0.000000
snp1278  2.644909e-01   1279       0.1968   2.3662    True     1.031421
snp1232  2.258332e-01   1233       0.1843   2.3312   False     0.000000
</pre>

<p>
FINEMAP picks SNPs in strong LD with the causal variant over the causal
variant:
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">query</span> = pip.sort_values(by=<span class="org-string">'log10bf'</span>, ascending=<span class="org-constant">False</span>).head().index
pd.DataFrame(x, columns=[<span class="org-string">'snp{}'</span>.<span class="org-builtin">format</span>(i) <span class="org-keyword">for</span> i <span class="org-keyword">in</span> <span class="org-builtin">range</span>(x.shape[1])])[query].corr()
</pre>
</div>

<pre class="example">
        snp1398   snp1404    snp676   snp1278   snp1232
snp1398  1.000000 -0.761444 -0.100979  0.793857  0.811527
snp1404 -0.761444  1.000000  0.047893 -0.582444 -0.598379
snp676  -0.100979  0.047893  1.000000 -0.107882 -0.106737
snp1278  0.793857 -0.582444 -0.107882  1.000000  0.973466
snp1232  0.811527 -0.598379 -0.106737  0.973466  1.000000
</pre>
</div>
</div>
</div>
<div id="postamble" class="status">
<p class="author">Author: Abhishek Sarkar</p>
<p class="date">Created: 2017-12-02 Sat 00:10</p>
<p class="validation"><a href="http://validator.w3.org/check?uri=referer">Validate</a></p>
</div>
</body>
</html>
