#+TITLE: PIP calibration
#+SETUPFILE: setup.org

* Setup                                                            :noexport:

  #+BEGIN_SRC emacs-lisp
    (org-babel-lob-ingest "/home/aksarkar/.emacs.d/org-templates/library.org")
  #+END_SRC

  #+RESULTS:
  : 1

  #+CALL: ipython3(venv="sse") :dir /scratch/midway2/aksarkar/sse

  #+RESULTS:
  : Submitted batch job 40808452

  #+BEGIN_SRC ipython
    import glob
    import matplotlib.pyplot as plt
    import pandas as pd
    import sse
  #+END_SRC

  #+RESULTS:
  :RESULTS:
  :END:

* Introduction

  Here we evaluate the concordance between posterior inclusion probability (PIP)
  computed using the Sum of Single Effects (SSE) variational approximation
  against other methods:

  - [[https://academic.oup.com/bioinformatics/article-lookup/doi/10.1093/bioinformatics/btw018][FINEMAP]] (2016)
    
  #+BEGIN_SRC ipython :results output
    !finemap
  #+END_SRC

  #+RESULTS:
  :RESULTS:
  
  |---------------------------------|
  | Welcome to FINEMAP v1.1         |
  |                                 |
  | (c) 2015 University of Helsinki |
  |                                 |
  | Help / Documentation:           |
  | - ./finemap --help              |
  | - www.christianbenner.com       |
  |                                 |
  | Contact:                        |
  | - christian.benner@helsinki.fi  |
  | - matti.pirinen@helsinki.fi     |
  |---------------------------------|
  
  Error: Option '--sss' or '--corr' must be supplied!
  
  :END:

  - [[http://www.sciencedirect.com/science/article/pii/S0002929716300957?via%253Dihub][DAP]] (2016)

  #+BEGIN_SRC sh :dir ~/.local/src/dap :eval never-export :exports both
    git describe --tags
  #+END_SRC

  #+RESULTS:
  : v1.0.0

  The simulation generative model is as follows:

  1. Sample a gene on chromosomes 1-22

    #+BEGIN_SRC ipython
      genes = (pd.read_table('/home/aksarkar/projects/singlecell-qtl/data/scqtl-genes.txt.gz')
               .set_index('gene')
               .query('source == "H. sapiens" and chr != "hsX" and chr != "hsY"'))
    #+END_SRC

    #+RESULTS:
    :RESULTS:
  :END:

  2. Extract genotypes of GEUVADIS individuals \(X\) from 100 kilobases
     upstream (respecting strand) of the gene start (~sse.evaluate.read_vcf~)

    #+BEGIN_SRC ipython
    genotype_files = ["/project/compbio/geuvadis/genotypes/GEUVADIS.chr{}.PH1PH2_465.IMPFRQFILT_BIALLELIC_PH.annotv2.genotypes.vcf.gz".format(i) for i in range(1, 23)]
  #+END_SRC

    #+RESULTS:
  :RESULTS:
  :END:

  3. [@3] Estimate MAF \(f_j\) from \(X\) (~sse.simulation.Simulation.estimate_maf~)
  4. Fix proportion of variance explained \(h^2 = 0.15\) (mean
     /cis/-heritability of gene expression across all genes; [[https://www.nature.com/articles/ng.3506][Gusev et al 2016]],
     [[https://www.nature.com/articles/ng.3506][Wheeler et al 2016]])
  5. Sample causal effects from \(\beta_j \sim N(0, 1)\) (~sse.simulation.Simulation.sample_effects~)
  6. Compute genetic variance \(V_g = \sum_j 2 * f_j * (1 - f_j) \beta_j^2\)
  7. Sample residuals \(\epsilon_i \sim N(0, V_g (1 / h^2 - 1))\)
  8. Compute phenotypes \(y_i = X_i\beta + \epsilon_i\) (~sse.simulation.Simulation.compute_liabilities~)

* One causal variant

  #+BEGIN_SRC ipython :results none
    one_causal = sse.evaluate.pip_calibration(
      genes=genes,
      genotype_files=genotype_files,
      num_genes=1,
      num_trials=10,
      num_effects=10,
      max_epochs=100)
  #+END_SRC

  #+RESULTS:
  :RESULTS:
  #+BEGIN_EXAMPLE
                          dap   finemap
    ENSG00000143641 0  0.930396  0.337642
                    1  0.980494  0.459722
                    2  0.861345  0.408443
                    3  0.942649  0.600707
                    4  0.964242  0.451974
                    5  0.987285  0.515028
                    6  0.922429  0.170489
                    7  0.927789  0.240316
                    8  0.990509  0.445197
                    9  0.970566  0.370578
  #+END_EXAMPLE
  :END:

  #+BEGIN_SRC ipython :ipyfile one-causal-variant-spearman.png
    plt.clf()
    plt.boxplot(one_causal.values, labels=one_causal.columns)
    plt.xlabel('Method')
    plt.ylabel('Spearman correlation of PIP')
    plt.gcf()
  #+END_SRC

  #+RESULTS:
  :RESULTS:
  [[file:one-causal-variant-spearman.png]]
  :END:
